File: word_distance_2.java
//TC: O(n)
//SC: O(n)
public class word_distance_2 {
    class WordDistance {

        Map<String,List<Integer>> map;
        
        public WordDistance(String[] wordsDict) {
            map = new HashMap<>();
            int idx = 0;
            for(String words : wordsDict) {
                map.computeIfAbsent(words, v->new ArrayList<Integer>()).add(idx++);
            }
        }
        
        public int shortest(String word1, String word2) {
            List<Integer> l1 = map.get(word1);
            List<Integer> l2 = map.get(word2);
            
            int i = 0, j = 0, minLen = Integer.MAX_VALUE;
           while(i < l1.size() && j < l2.size()) {
             minLen =  Math.min(minLen,Math.abs(l1.get(i) - l2.get(j)));
             if(l1.get(i) < l2.get(j)) i++;
             else j++;        
           }
        return minLen;
        }
    }
}

File: word_distance_2.py
# Time Complexity: O(n)
# Space Complexity: O(1)

class WordDistance:

    def __init__(self, wordsDict: List[str]):
        self.words = wordsDict
        self.map = defaultdict(list)
        for i, word in enumerate(self.words):
            self.map[word].append(i)

    def shortest(self, word1: str, word2: str) -> int:
        list1 = self.map[word1]
        list2 = self.map[word2]
        p1 = p2 = 0
        profit = 2**32
        # print(list1, list2)
        while p1 < len(list1) and p2 < len(list2):
            profit = min(profit, abs(list1[p1]-list2[p2]))
            if list1[p1] < list2[p2]:
                p1 += 1
            else:
                p2 += 1
        return profit

# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(wordsDict)
# param_1 = obj.shortest(word1,word2)
